### PR TITLE
[stable/fluent-bit]: fix extras indentation and making input.tail.Skip_Long_Lines configurable, add option to disable kubernetes annotations

### PR DESCRIPTION
**Pull request by varun-da**
#### What this PR does / why we need it:
There is a bug with identation to add extra outputs, inputs and filters. The indentation should be set to 4, currently it is set to 8 and results in the following error:
```
Fluent-Bit v0.14.6
Copyright (C) Treasure Data

[2018/11/26 18:04:02] [  Error] File /fluent-bit/etc/fluent-bit.conf
[2018/11/26 18:04:02] [  Error] Error in line 23: Each key must have a value
```

And making input tail plugin's `Skip_Long_Lines` option configurable, https://docs.fluentbit.io/manual/input/tail#config. Currently it cannot be changed and is set to On. I would like to configure it to be turned Off. By default it is coded to be turned on. 

And adding an option to disable kubernetes annotations, based on
https://github.com/fluent/fluent-bit/blob/master/plugins/filter_kubernetes/kube_conf.c#L54 and
https://github.com/fluent/fluent-bit/issues/579#issuecomment-386378519. This was tested on my kubernetes cluster and works.

Disabling of kubernetes annotations is very useful, because for some implementations like [datadog autodiscovery](https://docs.datadoghq.com/agent/autodiscovery/?tab=docker), the config is included in the annotations and when annotations are injected by default into each log line, this happens:
```
{"log":"17:01:45,646 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [/etc/logback-standard.xml] at [file:/etc/logback-standard.xml]\n","stream":"stdout","time":"2018-11-26T17:01:45.708131872Z","kubernetes":{"pod_name":"reference-app-5b9545d86c-hm4z4","namespace_name":"reference-app","pod_id":"eb9c2c79-f19c-11e8-99ac-1208343f026a","labels":{"app":"reference-app","chart":"da-generic-0.20.0","heritage":"Tiller","pod-template-hash":"1651018427","release":"reference-app"},"annotations":{"ad.datadoghq.com/reference-app.check_names":"[\\\"reference-app\\\"]","ad.datadoghq.com/reference-app.init_configs":"[{\\\"is_jmx\\\":true,\\\"collect_default_metrics\\\":true,\\\"conf\\\":[{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"OperatingSystem\\\"}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"Copy\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"PS Scavenge\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"ParNew\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"G1 Young Generation\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.minor_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"MarkSweepCompact\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"PS MarkSweep\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"ConcurrentMarkSweep\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"GarbageCollector\\\",\\\"name\\\":\\\"G1 Mixed Generation\\\",\\\"attribute\\\":{\\\"CollectionCount\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_count\\\"},\\\"CollectionTime\\\":{\\\"metric_type\\\":\\\"gauge\\\",\\\"alias\\\":\\\"jmx.gc.major_collection_time\\\"}}}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"Memory\\\",\\\"attribute\\\":[\\\"HeapMemoryUsage\\\",\\\"NonHeapMemoryUsage\\\"]}},{\\\"include\\\":{\\\"domain\\\":\\\"java.lang\\\",\\\"type\\\":\\\"OperatingSystem\\\",\\\"attribute\\\":[\\\"ProcessCpuLoad\\\"]}}}]}]","ad.datadoghq.com/reference-app.instances":"[{\\\"host\\\": \\\"%%host%%\\\", \\\"port\\\": 7199, \\\"tags\\\": [\\\"reference-app\\\"]}]"},"host":"ip-0-0-0-0.ec2.internal","container_name":"reference-app","docker_id":"94fe42"}}
```

#### Special notes for your reviewer:
bumped the app to latest release version 0.14.8, https://fluentbit.io/announcements/v0.14.8/, https://github.com/fluent/fluent-bit/releases/tag/v0.14.8

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by varun-da** [comment]
Hi @edsiper, @kfox1111, @davidkarlsen and @unguiculus, Can you please review these changes and let me know if any changes are requested. I need the extra filters to work as annoations in logs are causing a lof of issues on our logging framework.

Thanks!

**Comment by kfox1111** [comment]
Someone from the charts team, please approve the indention fix. this is now the 5th pr trying to fix it. Where is everyone?

**Comment by varun-da** [comment]
@prydonius @mgoodness @linki @davidkarlsen @sameersbn @technosophos @bacongobbler @fibonacci1729 @gabrtv @rimusz @scottrigby @slack any chance I can have this PR reviewed and possibly approved and merged (if no changes are requested)? Otherwise will have to clone to chart to my repo and maintain a local copy. 

My apologies for tagging all of you. I am trying to avoid duplicating the chart. I have already done it for coredns a week prior to this week, PR was not approved for months.

Thanks!

**Review by bacongobbler — COMMENTED** [review]
This PR was submitted less than 24 hours ago. Please have a little more patience before pinging everyone and their cat (including people who are not chart maintainers like myself) :)

The changes in this PR look good to me, but I'll defer to the chart maintainers for review/approval.

**Comment by varun-da** [comment]
Hi @bacongobbler, thanks for responding. I am confused, @kfox1111 is listed as a Chart maintainer: https://github.com/helm/charts/blob/master/stable/fluent-bit/Chart.yaml#L15-L16. But he cannot approve?

Thanks!

**Comment by kfox1111** [comment]
@bacongobbler This pr was submitted 24 hours ago. Others have not.  I was waiting on one fix for 2 months while 5 pr's piled up. It did look like it merged and it didn't ping me. sorry for the noise. 2 months is a very long time though. How do we help increase the reviews? 
see https://github.com/helm/charts/issues/7632 for a list of pr's that all hit the same issue.
I had to even beg for help on slack with no replies. I had some contributors just give up and close useful pr's.

I've got some more charts to contribute too, and I've kind of given up on contributing them until things get better. How do we move forward? Should I start adding myself to owners files? Can I get a reviewer bit?

Thanks,
Kevin

**Comment by varun-da** [comment]
@kfox1111 As you are the maintainer for this PR, I think you can just a post message approve (if you think this is good) and then perhaps @bacongobbler can merge this.

File: stable/fluent-bit/templates/config.yaml [hunk 1]
```diff
@@ -31,19 +31,20 @@ data:
...
         Refresh_Interval 5
         Mem_Buf_Limit    {{ .Values.input.tail.memBufLimit }}
-        Skip_Long_Lines  On
+        Skip_Long_Lines  {{ .Values.input.tail.skipLongLines }}
 {{- if .Values.trackOffsets }}
         DB               /tail-db/tail-containers-state.db
         DB.Sync          Normal
 {{- end }}
-{{ .Values.extraInputs | indent 8 }}
+{{ .Values.extraInputs | indent 4 }}
```
Comment (kfox1111) [inline stable/fluent-bit/templates/config.yaml:39]: Why is this here? This will break things.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: Infact, there was a recent patch, which I was confusing this one with, that merged fairly recently:
https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: @kfox1111 this is similar to `.Values.extraFilters` and `.Values.extraOutputs` to add additional [INPUT] section.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: [9ab3cbb](https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506) is incorrect because it is trying to add additional parameters/options to the existing plugin. my understanding is `extraInputs` is used for adding additional `[INPUT]` sections and so on with `extraOutputs` and `extraFilters`. Maybe a separate variable is needed to implement [9ab3cbb](https://github.com/helm/charts/commit/9ab3cbbc28bf0416cf5400fb9ea4679763807506).

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: @kfox1111 looking at the readme, it is clear that those variables are there for adding additional sections, not adding config values to existing sections:


extraInputs | Add extra Input sections to config | ``
-- | -- | --
extraFilters | Add extra Filter sections to config | ``
extraOutputs | Add extra Output sections to config | ``

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: yeah, I think its correct in that the idea was to let the user add more options to the existing input/filter/outputs.

I don't think toml lets you have multiple sections with the same name? so it won't ever work I think.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: Yeah, I think that readme could probably be worded better. its sections as in lines of text, not sections as in a new block of toml. Any idea how to more concisely wordsmith that?

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: Oh yes, it does,  please see the sections doc here, https://docs.fluentbit.io/manual/configuration/schema#sections, 

> Multiples sections can exists on the same file.

And a example of this filter sections: https://docs.fluentbit.io/manual/filter/nest#configuration-file-3

thanks!

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: I think the readme is correct, if you take the `extraOutputs` it is added at the end of all the outputs sections, so this would mean that only `http` output can have additional parameters and the other output sections file, forward, es and splunk cannot. 

also the extras{Inputs,Ouputs,Filters} were added as part of https://github.com/helm/charts/pull/8600, which clearly states the intent was to add additional sections. It was misused and then an attempt to patch in 9ab3cbb.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: oh gosh... that feels really weird... only sax style parsers would be able to handle it. Dom parsers will be inconsistent or break. this parser chokes on that example:
http://toml-online-parser.ovonick.com/

Ok, well, if its a design feature of fluent-bit, then it is what it is. :/

Do the sections all need to be in order? If not, I'd recommend leaving these options alone, and maybe adding an .extraSections option towards the bottom?

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: Also the author of that PR admits it was not intended for that purpose and the fix was not correct: https://github.com/helm/charts/pull/8825#issuecomment-435867527, but it was never reverted. and another PR has been opened to fix the same issue: https://github.com/helm/charts/pull/9008.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: I am not sure if the sections have to be in order, docs don't mention it, https://docs.fluentbit.io/manual/configuration/schema#sections. So I am not sure.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: also this comment addresses how the same variable can be used to have a section and a parameter for the section above if needed: https://github.com/helm/charts/pull/9008#issuecomment-439481904.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: hmm.. yeah. that pr does seem to have intended it to be for extra sections... I never reviewed it though. I think I would have likely blocked it in that form, asking the same questions I am now...

shoehorning yaml into toml like this is really kind of gross. I'm kind of of the opinion, make the 20% of the config options that 80% of the users need super easy to set and for the rest, just let the user upload their own fully customized configmap via existingConfigMap. Otherwise we're going to keep adding mapping after mapping of stuff from yaml to toml. Have we reached that point here?

I get making Skip_Long_Lines configurable since it can't be overridden otherwise.

Multiple INPUTS/FILTERS/OUTPUT sections feels like its over the line though?

**Comment by bacongobbler** [comment]
> and then perhaps @bacongobbler can merge this.

I am refraining from merging this PR because I am [not on the OWNERS](https://github.com/helm/charts/blob/master/OWNERS) list. I have org-level access, but as I mentioned before, I am not a chart maintainer.

File: stable/fluent-bit/templates/config.yaml (hunk 1 above)
Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: Well, the thing is you have to take into consideration the application, in this case fluent-bit, I really only want to (in most cases) change the INPUT/FILTERS/OUTPUT sections. And change the logs to a format I wanted and send it to a backend. 

Now if I have to overwrite a configmap, I to write the whole config file for fluent-bit and then apply it. 

Right now with this simple config, it works.

```
fluent-bit:
  backend:
    forward:
      host: logging
      port: 24224
  extraFilters: |-

    [FILTER]
        Name nest
        Match kube.*
        Operation lift
        Nested_under kubernetes
        Add_prefix kubernetes_
```

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: https://github.com/helm/charts/pull/3573

Also the implementation of overriding configmap is done weirdly, so I have to have a separate process create the configmap and then have that configmap name provided in the chart values file. And that configmap is out of scope for helm to manage, update, etc. I think the configmap override should have been done at the `data` key at https://github.com/helm/charts/blob/master/stable/fluent-bit/templates/config.yaml#L11, so I could have provided the configmap in the values file. I do not want to manage kubernetes objection creation separately, helm and charts are made to do exactly that.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: The issue I have is that to add extra options to say the existing filter, then you have to know how much whitespace is involved which should be an implementation detail that should never be exposed.

I think your example could be done like:
```yaml
fluent-bit:
  backend:
    forward:
      host: logging
      port: 24224
  extraSections: |-

    [FILTER]
        Name nest
        Match kube.*
        Operation lift
        Nested_under kubernetes
        Add_prefix kubernetes_
```

if this were added at the end of the configmap:
```
{{ .Values.extraSections | indent 4 }}
```
It would allow adding to the existing values without needing to know the implementation detail of how far indented the config needs to be, I'd be ok with that. Could you test that?

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: sure, testing now.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: Just going by this example, filters might have to be in sequence: https://docs.fluentbit.io/manual/filter/nest#configuration-file-3

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: It might work in my test, because I am just adding filters, but it might break for others, who need the correct sequence.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: @edsiper can you please weigh in on if the config sections will always work out of order?

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: In general, there's a reason why configmaps are text blobs in k8s and not yaml or something. The k8s developers realized a common problem in config management is that config managers try and make the config standardized into some format (say yaml) and then map it into the application's config. Since there is a million ways to write out configs, there is a million ways to convert the input to the output. This leads to a snowballing of config management code, unit tests, etc. Its a painful treadmill. k8s put this out of scope and said, if you want to pass config to sendmail in m4 macro's, then stick it in a configmap and we'll ship it without touching it. It will just work. Helm suffers from the config mapping issue quite a bit. The more config options you want to have in values, the more conditional/gotl template crazyness goes on until the template is unreadable. By allowing the user to upload a configmap directly, that gets us off the treadmill of supporting every possible option in every possible combination. If you fit in the 80% common case, values is easy and there. If you don't you upload your config and it supports every possible option you might need, without the chart becoming unreadable/unmaintainable. This feels like a good tradeoff of developer and user time.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: As it stands, I can't even do that, the override configmap options is for the configmap to be created separately, outside of the helm chart entirely.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: Yup. Do:
```
kubectl create configmap fluent-bit --from-file=fluent-bit.conf
```

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: adding the sections at the end worked for me, but not certain if it will work in all scenarios. doing a test with nesting and putting the filters out of order to see if that works as intended.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: Yeah, well I understand how to make a configmap from a file. That is not at all the issue, it would break the automation we have, which expects logging to be installed via charts entirely.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: out order filters break the flow though. in this example: https://docs.fluentbit.io/manual/filter/nest#configuration-file-3, which I implemented a similar one, if they are not in the correct sequence, it breaks it, i.e. the nesting and then the following filtering, key removal, etc, all don't work.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: thanks for all the help @kfox1111! i have a tight timeline. i am going to import this chart into the internal repo. if you want i will leave this PR open, otherwise will go ahead and close it. thank you!

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: Another option is to throw the configmap into your own chart. and possibly add the fluent-bit one as a subchart. At some point, you gotta maintain your own things. But this allows you to share as much as possible. If you have to do absolutely everything through helm its likely you will need an integration chart like that at some point.

I'm really worried about the current templates getting unmaintainable. We've broken them lots of times already adding options. The more we add, the more likely it is to be broken again.

**Comment by varun-da** [comment]
closing as this will not be approved.

File: stable/fluent-bit/templates/config.yaml (hunk 1 above)
Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: FYI, i was talking with donbowman and he mentioned actually, all of the following are potentially valid and different configs:
 * [input] [input] [filter] [filter] [output] [output]
 * [input] [filter] [output] [input] [filter] [output]
and even orderings of the filters can be different with different results...
 * [input] [filter a] [filter b] [output]
 * [input] [filter b] [filter a] [output]

So, extraInputs/extraOutputs doesn't really make too much sense at 4 indentions.  extraFilters kind of does, but you may want a preExtraFilters too then, and you still might want an extraSections.

All this to say, its super complicated making the values flexible enough to cover all cases. Instead, lets just cover the [input][filter][output] use case, and use your own configmap if you must differ.

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: ok cool. I used the approach you suggested and went with installing the configmap using a parent/bridge chart (which we already had) and fluent-bit as a dependency. worked out nicely! thanks @kfox1111!

Comment (varun-da) [reply stable/fluent-bit/templates/config.yaml:39]: yes, they are all valid, filters one has to be careful of, because like you stated and with my testing, the sequence matters to get the results one wants.

Comment (kfox1111) [reply stable/fluent-bit/templates/config.yaml:39]: Ok, cool. glad its working for you. :)

Comment (mattparkes) [reply stable/fluent-bit/templates/config.yaml:39]: The key is called `extraInputs`. I wanted to add _extra inputs_ to my configmap, so I tried this key and it didn't work!

Also, `indent 4` is far more flexible than `indent 8` because I can self indent 4 to achieve indent 8 if I want to add entries to an existing INPUT section, but I cant negatively indent to add an entire new INPUT section.
